### ISSUE TITLE
Jackson @JsonAnyGetter emits Invalid JSON for a POJO 

### ISSUE THREAD

**Issue by Bishnup1995** [FasterXML/jackson-databind#4091]
This issue occurs when a key in the Map has the same name as a class member in the POJO. Find the test code below:

```
static class BeanWithProperty {
        private String b;
        final static Map<String, Boolean> extra = new HashMap<String, Boolean>();

        static {
            extra.put("a", Boolean.TRUE);
            extra.put("b", Boolean.TRUE);
        }

        public String getB() {
            return b;
        }

        public void setB(String b) {
            this.b = b;
        }

        @JsonAnyGetter
        public Map<String, Boolean> getExtra() {
            return extra;
        }
    }

    public void testAnyGetterWithDuplicateProperty() throws Exception {
        String b = "false";
        BeanWithProperty beanWithProperty = new BeanWithProperty();
        beanWithProperty.setB("false");
        ObjectMapper objectMapper = new ObjectMapper();
        objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
        String json = objectMapper.writeValueAsString(beanWithProperty);
        BeanWithProperty beanWithProperty1 = objectMapper.readValue(json, BeanWithProperty.class);
        // should not fail
        assertEquals(beanWithProperty1.getB(), b);
    }
```

**The test case fails**

I am observing same behavior in all the versions.

**Comment by yawkat** [comment]
jackson will not detect all written duplicates unless you enable STRICT_DUPLICATE_DETECTION

**Comment by Bishnup1995** [comment]
@yawkat Tried with `STRICT_DUPLICATE_DETECTION` but no luck. Please find the test code below:

```
public void testAnyGetterWithDuplicateProperty() throws Exception {
        String b = "false";
        BeanWithProperty beanWithProperty = new BeanWithProperty();
        beanWithProperty.setB("false");
        ObjectMapper objectMapper = new ObjectMapper();
        objectMapper.configure(DeserializationFeature.FAIL_ON_UNKNOWN_PROPERTIES, false);
        objectMapper.enable(JsonParser.Feature.STRICT_DUPLICATE_DETECTION);
        String json = objectMapper.writeValueAsString(beanWithProperty);
        System.out.println(json);
        // should not fail
        BeanWithProperty beanWithProperty1 = objectMapper.readValue(json, BeanWithProperty.class);
        assertEquals(beanWithProperty1.getB(), b);
    }
```

**Comment by yawkat** [comment]
you have to enable it for the generator

**Comment by Bishnup1995** [comment]
@yawkat can you add a reference example ?

**Comment by yawkat** [comment]
`JsonGenerator.Feature.STRICT_DUPLICATE_DETECTION`

**Comment by Bishnup1995** [comment]
@yawkat Thank you for your response. I am getting the exception when trying to read the string. The issue I want to highlight  here is invalid JSON string when using `objectMapper.writeValueAsString(beanWithProperty)`. I tried printing the jsonString and below is the output:
`{"b":"false","a":true,"b":true}`

**Comment by yawkat** [comment]
when you enable the generator failure you will get an exception when trying to serialize this bean

**Comment by Bishnup1995** [comment]
> when you enable the generator failure you will get an exception when trying to serialize this bean

Yes correct. I get that. The issue I want to highlight here is invalid JSON string when using `objectMapper.writeValueAsString(beanWithProperty)`. I tried printing the jsonString and below is the output:
`{"b":"false","a":true,"b":true}`

**Comment by yawkat** [comment]
It is intentional that not all cases where duplicate json keys are generated lead to an exception by default. This has performance reasons, but duplicate properties are technically not even invalid. If you want to forbid generation of duplicate keys, you must enable `JsonGenerator.Feature.STRICT_DUPLICATE_DETECTION`. Duplicate keys without this flag enabled are expected and imo should not be addressed unless you can guarantee no performance impact.

**Comment by Bishnup1995** [comment]
@yawkat IMHO generating a wrong JSON string may lead to in correct result. We should avoid doing that. @cowtowncoder Could you add your opinion ?

**Comment by cowtowncoder** [comment]
@Bishnup1995 checking for duplicates adds measurable overhead so by default such checking should not be imposed. However, it'd be nice to have counterpart to `StreamReadFeature.STRICT_DUPLICATE_DETECTION` (like `StreamWriteFeature.STRICT_DUPLICATE_DETECTION`).

It would also be nice to be able to avoid emitting duplicate values for `@JsonAnyGetter` but this is not easy to implement because nothing in `JsonGenerator` keeps track of what has been output -- and the order in which properties are written can also vary ("any getter" could be output before or after regular properties).

Of above, I think former would be easier to implement; if anyone has time and interest, I'd be happy to help. But I don't think this feature is something I would work on near term, due to everything else that needs attention.

In short term you will need to ensure not to expose same properties via any getter as regular properties.

**Comment by Bishnup1995** [comment]
@cowtowncoder I have created a pull request for a possible solution. Can you take a look?

**Comment by cowtowncoder** [comment]
Had a look; unfortunately I do not think that approach can be made to work either correctly or efficiently.
